File: src/integration/spring/src/jvmTest/java/community/flock/wirespec/integration/spring/java/application/Controller.java
package community.flock.wirespec.integration.spring.java.application;

import community.flock.wirespec.integration.spring.java.generated.endpoint.*;
import community.flock.wirespec.integration.spring.java.generated.model.ApiResponse;
import community.flock.wirespec.integration.spring.java.generated.model.Pet;
import org.springframework.web.bind.annotation.RestController;

import java.util.Collections;
import java.util.Optional;
import java.util.concurrent.CompletableFuture;

@RestController
public class Controller implements 
    AddPet.Handler,
    GetPetById.Handler,
    UpdatePet.Handler,
    DeletePet.Handler,
    FindPetsByTags.Handler,
    UploadFile.Handler {

    private final Service service;

    public Controller(Service service) {
        this.service = service;
    }

    @Override
    public CompletableFuture<AddPet.Response<?>> addPet(AddPet.Request request) {
        service.create(request.body());
        return CompletableFuture.completedFuture(new AddPet.Response200(Optional.of(100), request.body()));
    }

    @Override
    public CompletableFuture<GetPetById.Response<?>> getPetById(GetPetById.Request request) {
        Optional<Pet> pet = service.list.stream()
                .filter(it -> it.id().isPresent() && it.id().get().equals(request.path().petId()))
                .findFirst();

        GetPetById.Response<?> response = pet
                .<GetPetById.Response<?>>map(GetPetById.Response200::new)
                .orElseGet(GetPetById.Response404::new);

        return CompletableFuture.completedFuture(response);
    }

    @Override
    public CompletableFuture<UpdatePet.Response<?>> updatePet(UpdatePet.Request request) {
        service.update(request.body());
        return CompletableFuture.completedFuture(new UpdatePet.Response200(request.body()));
    }

    @Override
    public CompletableFuture<DeletePet.Response<?>> deletePet(DeletePet.Request request) {
        long id = 1L;
        service.delete(id);
        return CompletableFuture.completedFuture(new DeletePet.Response400());
    }

    @Override
    public CompletableFuture<FindPetsByTags.Response<?>> findPetsByTags(FindPetsByTags.Request request) {
        return CompletableFuture.completedFuture(new FindPetsByTags.Response200(Collections.emptyList()));
    }

    @Override
    public CompletableFuture<UploadFile.Response<?>> uploadFile(UploadFile.Request request) {
        Optional<byte[]> file = request.body().file();
        if (file.isEmpty()) throw new RuntimeException("Missing file");
        service.upload(file.get());
        return CompletableFuture.completedFuture(
            new UploadFile.Response200(
                new ApiResponse(
                    Optional.of(200),
                    Optional.of("type"),
                    Optional.of(request.body().additionalMetadata().orElse("none"))
                )
            )
        );
    }
}

File: src/integration/spring/src/jvmTest/kotlin/community/flock/wirespec/integration/spring/kotlin/application/Controller.kt
package community.flock.wirespec.integration.spring.kotlin.application

import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.AddPet
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.DeletePet
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.FindPetsByTags
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.GetPetById
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.RequestParrot
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.UpdatePet
import community.flock.wirespec.integration.spring.kotlin.generated.endpoint.UploadFile
import community.flock.wirespec.integration.spring.kotlin.generated.model.ApiResponse
import org.springframework.web.bind.annotation.RestController

@RestController
class Controller(
    private val service: Service,
) : AddPet.Handler,
    GetPetById.Handler,
    UpdatePet.Handler,
    DeletePet.Handler,
    FindPetsByTags.Handler,
    UploadFile.Handler,
    RequestParrot.Handler {

    override suspend fun addPet(request: AddPet.Request): AddPet.Response<*> {
        service.create(request.body)
        return AddPet.Response200(request.body, 200)
    }

    override suspend fun getPetById(request: GetPetById.Request): GetPetById.Response<*> = service.list.find { it.id == request.path.petId }
        ?.let { GetPetById.Response200(it) }
        ?: GetPetById.Response404(Unit)

    override suspend fun updatePet(request: UpdatePet.Request): UpdatePet.Response<*> {
        service.update(request.body)
        return UpdatePet.Response200(request.body)
    }

    override suspend fun deletePet(request: DeletePet.Request): DeletePet.Response<*> {
        val id = 1L
        return service.delete(id).let {
            DeletePet.Response400(Unit)
        }
    }

    override suspend fun findPetsByTags(request: FindPetsByTags.Request): FindPetsByTags.Response<*> = FindPetsByTags.Response200(emptyList())

    override suspend fun uploadFile(request: UploadFile.Request): UploadFile.Response<*> {
        val file = request.body.file ?: error("Missing file")
        service.upload(file)
        return UploadFile.Response200(
            ApiResponse(
                code = 200,
                type = "type",
                message = request.body.additionalMetadata ?: "none",
            ),
        )
    }

    override suspend fun requestParrot(request: RequestParrot.Request): RequestParrot.Response<*> = RequestParrot.Response200(
        body = request.body,
        XRequestID = request.headers.XRequestID,
        RanDoMHeADer = request.headers.RanDoMHeADer,
        QueryParamParrot = request.queries.QueryParam,
        RanDoMQueRYParrot = request.queries.RanDoMQueRY,
    )
}
